api/service.go

```diff
@@ -524,6 +528,158 @@ func (s *service) RemoveAppToken(ctx context.Context, req *pb.RemoveAppTokenRequ
+	}, nil
+}
+
+// GetFolder handles a get folder request.
+func (s *service) GetFolder(ctx context.Context, req *pb.GetFolderRequest) (*pb.GetFolderReply, error) {
+	log.Debugf("received get folder request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	folder, err := s.getFolderWithScope(ctx, req.Name, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := s.ipfsClient.Unixfs().Get(ctx, path.New(folder.Path))
+	if err != nil {
+		return nil, err
+	}
+	defer node.Close()
+
+	return folderToPbFolder(folder, node)
+}
+
+func folderToPbFolder(folder *c.Folder, node ipfsfiles.Node) (*pb.GetFolderReply, error) {
+	entries, err := dirToPbSlice(node)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.GetFolderReply{
+		ID:        folder.ID,
+		Path:      folder.Path,
+		Name:      folder.Name,
+		Public:    folder.Public,
+		ProjectID: folder.ProjectID,
+		Created:   folder.Created,
+		Entries:   entries,
+	}, nil
+}
+
+func dirToPbSlice(node ipfsfiles.Node) (entries []*pb.GetFolderReply_Entry, err error) {
+	err = ipfsfiles.Walk(node, func(path string, n ipfsfiles.Node) error {
+		defer n.Close()
+		if path == "" { // This is the root
+			return nil
+		}
+		size, err := n.Size()
+		if err != nil {
+			return err
+		}
+		entry := &pb.GetFolderReply_Entry{
+			Path: path,
+			Size: size,
+		}
+		switch n.(type) {
+		case ipfsfiles.Directory:
+			entry.IsDir = true
+		}
+		entries = append(entries, entry)
+		return nil
+	})
+	return
+}
+
+// ListFolders handles a list folders request.
+func (s *service) ListFolders(ctx context.Context, req *pb.ListFoldersRequest) (*pb.ListFoldersReply, error) {
+	log.Debugf("received list folders request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	proj, err := s.getProjectForScope(ctx, req.ProjectID, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	folders, err := s.collections.Folders.List(ctx, proj.ID)
+	if err != nil {
+		return nil, err
+	}
+	list := make([]*pb.GetFolderReply, len(folders))
+	for i, folder := range folders {
+		node, err := s.ipfsClient.Unixfs().Get(ctx, path.New(folder.Path))
+		if err != nil {
+			return nil, err
+		}
+		list[i], err = folderToPbFolder(folder, node)
+		if err != nil {
+			return nil, err
+		}
+		node.Close()
+	}
+
+	return &pb.ListFoldersReply{List: list}, nil
+}
+
+// RemoveFolder handles a remove folder request.
+func (s *service) RemoveFolder(ctx context.Context, req *pb.RemoveFolderRequest) (*pb.RemoveFolderReply, error) {
+	log.Debugf("received remove folder request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	folder, err := s.getFolderWithScope(ctx, req.Name, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = s.collections.Folders.Delete(ctx, folder.ID); err != nil {
+		return nil, err
+	}
+
+	if err = s.ipfsClient.Pin().Rm(ctx, path.New(folder.Path)); err != nil {
```

Comment:
There's only one pin at the root that needs to be removed.

Reply:
Does unpinning also unpin _orphaned_ children of the unpinned cid?

Reply:
The children don't ever get pinned. There could be a cross folder file that is under two different pins from different roots. The second one would remain as expected.

Reply:
Yep, actually I tried to exclude that case with the _ophaned_ word (which I don't know if is correct, took it from ORM worlds :P).

If a child is orphaned (no root links to it), I guess should be unpinned. I guess this indirectly will behave as how IPFS understands unpinning hierarchies, so I guess its ok.

Reply:
Yep, they have this concept of a "recursive" type pin which is used to unpin as well.